[블랙잭 2단계] 다이스 미션 제출합니다.

moondev03 opened the pull request:
# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

<!--
학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
-->

## BlackJack View Interface
BlackJackInputView / BlackJackOutputView => Interface
BlackJack Interface로 추가하여 어떤 컨트롤러든 BlackJackGame을 사용할 수 있도록 했습니다.

<br>

## PersonCards
Hand에 있던 카드 목록을 따로 객체로 분리했습니다.
```kotlin
class PersonCards {
    private val cards: MutableList<Card> = mutableListOf()

    fun add(card: Card) {
        cards.add(card)
    }

    fun list(): List<Card> {
        return cards.toList()
    }
}
```

<br>

## GameResult
기존 승패 여부 계산 및 카운트를 진행하던 객체를 2단계 요구사항에 맞춰
플레이어/딜러의 수익 계산을 위한 객체로 수정하였습니다.

<br>

## 코드 리뷰 커뮤니케이션

<!--
리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
-->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

<!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

<!-- 필요하다면 스크린샷을 첨부하세요. -->

<img width="387" alt="image" src="https://github.com/user-attachments/assets/66bae1f9-b794-4678-a2f3-333f27840495" />

moondev03 committed:
test: Test Coverage 참고하여 Deck 테스트 수정

- 덱 생성 시 크기에 대한 검증을 하도록 수정

moondev03 committed:
docs(README): 2단계 기능 구현 목록 추가 작성

moondev03 committed:
chore: HitFlag -> IsHit 네이밍 수정

moondev03 committed:
feat: 플레이어 배팅 기능 구현

moondev03 committed:
test: ResultStateTest 코드 작성

moondev03 committed:
refactor: View 관련 확장 함수 파일 분리

moondev03 committed:
test: PersonState BlackJack 상태 추가

moondev03 committed:
feat: GameResult 기능 변경

- 승패 카운트 계산 기능 제거
- 배당 계산 기능 구현

moondev03 committed:
test: BlackJackGame !! 사용 제거

moondev03 committed:
refactor: Hand 책임 분리

- PersonCards 객체 추가
- 수정 사항 반영

moondev03 committed:
docs(README): 기능 구현 목록 갱신

moondev03 committed:
refactor: 배팅 금액 객체 추가

- 배팅 금액 검증 기능 구현
- retryWhenException 기능 구현

moondev03 committed:
refactor: 게임 결과 플레이어 손실 계산 기능 구현

moondev03 committed:
refactor: BlackJack View Interface 구현

Gyuil-Hwnag commented on README.md:
```diff
@@ -33,4 +33,44 @@
+
+<br>
+
+## Level 2 - 블랙잭 (베팅)
+
+블랙잭 게임을 변형한 프로그램을 구현한다. <br>
+블랙잭 게임은 딜러와 플레이어 중 카드의 합이 21 또는 21에 가장 가까운 숫자를 가지는 쪽이 이기는 게임이다.
```
2단계에 대한 내용 작성좋네요👍

Gyuil-Hwnag commented on src/main/kotlin/blackjack/domain/BetAmount.kt:
```diff
@@ -0,0 +1,20 @@
+        require(value >= MIN_BET_AMOUNT) { INVALID_BET_AMOUNT_MESSAGE }
+    }
+
+    fun toInt(): Int = value
+
+    fun toDouble(): Double = value.toDouble()
```
해당 내용들이 들어간 이유가 궁금합니다.

moondev03 replied:
`betAmount.value.toDouble()` 대신 `betAmount.toDouble()`을 사용하면, 
`value`에 직접 접근할 필요 없이 더 직관적으로 값을 변환할 수 있다고 생각했습니다.
  
또한, `value`를 `private`으로 감추면서 캡슐화를 유지할 수 있어, 
향후 내부 구현을 변경하더라도 외부 코드에 영향을 주지 않는 장점이 있다고 생각합니다!

Gyuil-Hwnag commented on src/main/kotlin/blackjack/domain/GameResult.kt:
```diff
@@ -2,15 +2,66 @@ package blackjack.domain
-    val winStatus: Map<Player, ResultState> =
-        players.associateWith { player ->
-            ResultState.calculateWin(player, dealer)
+    val playerPayouts: Map<Player, Double>
+    val dealerProfit: Double
+
+    init {
+        var totalDealerProfit = 0.0
+        playerPayouts =
+            players.associateWith { player ->
+                val resultState = ResultState.calculateWin(player, dealer)
+                val payout = calculatePayout(player, resultState)
+                totalDealerProfit += calculateDealerProfit(player, resultState)
+                payout
+            }
+        dealerProfit = totalDealerProfit
+    }
```
init 에서 변수를 지정하도록 하신 이유가 있을까요?

moondev03 replied:
객체가 생성될 때 init 블록에서 이 값을 계산하고 초기화하여, 
이후에 객체를 사용할 때 바로 결과를 참조할 수 있게 하기 위해 이렇게 구현하였습니다.

Gyuil-Hwnag commented on src/main/kotlin/blackjack/domain/GameResult.kt:
```diff
@@ -2,15 +2,66 @@ package blackjack.domain
+import blackjack.domain.state.PersonState
 import blackjack.domain.state.ResultState
 
 class GameResult(dealer: Dealer, players: List<Player>) {
-    val winStatus: Map<Player, ResultState> =
```
GameResult 가 배팅금액과 이윤을 계산하는 객체로 바뀌었군요.

Gyuil-Hwnag commented on src/main/kotlin/blackjack/domain/person/PersonCards.kt:
```diff
@@ -0,0 +1,15 @@
+
+import blackjack.domain.card.Card
+
+class PersonCards {
+    private val cards: MutableList<Card> = mutableListOf()
```
Hand라는 객체가 있는데, PersonCards 객체가 추가된 이유가 있을까요?

moondev03 replied:
저도 고민했던 부분인데 기존 `Hand`는 카드만을 관리하기 위해 구현했으나, 
리팩토링을 수행하면서 점수 객체가 포함되게 되었습니다.
이로 인해 `Hand`의 역할이 늘었났다고 판단해서 분리하고 싶었습니다.

책임을 분리한 후 각각의 모델은 다음과 같은 책임을 수행하게 되었습니다.
- `PersonCards`는 단지 카드를 관리하는 객체로서, 카드 추가나 카드 목록을 조회하는 책임만 집니다. 
- `Hand`는 `PersonCards`가 제공하는 카드 목록을 사용하여 점수 계산과 같은 역할을 수행합니다. 

이 구조로 변경하므로써 두 모델의 하나의 책임을 가지게 되었다고 생각합니다!

Gyuil-Hwnag replied:
그렇군요.
살짝 찜짬한 의문이 들었던 것은 아래와 같은 뎁스 형태로 구현이 되어 있어서 였습니다.
`Player -> Hand -> PersonCards`
이렇게 객체간의 뎁스 형태로 되어 있다보면, 추후 요구사항이 변경되거나 뎁스 하단의 구조가 변경이 된다면, 자칫 잘못하면 전체 구조가 바뀔 수 있을 수 있겠다는 생각이 듭니다.
뎁스 구조에서 `점수 계산기 / 카드 관리`이렇게 분리를 해서 최 상단에서 계산을 해보도록 하는 것은 어떤가 라는 생각이 들었어요!💪

(여러가지 방법을 구현을 해보면서, 각 방법의 장단점을 파악하는 것에서 얻는 것이 있을 것 같다는 생각이 들어서 위에서 말한대로 시간이 남는만큼 한번 구조를 개선해봐도 좋을 것 같습니다)

moondev03 replied:
다시 PersonCards를 제거하고 Hand에서 관리하도록 변경했습니다.

cf1fd6042cd59ce4faca31c929a2d6eb07d6a071

Gyuil-Hwnag commented on src/main/kotlin/blackjack/domain/state/PersonState.kt:
```diff
@@ -8,16 +8,18 @@ enum class PersonState(val isFinal: Boolean) {
             val isDealer = person is Dealer
 
             return when {
+                person.cards().size == 2 && person.score() == BLACKJACK_SCORE -> BLACKJACK
```
블랙잭 규칙이 추가되었군요👍

Gyuil-Hwnag commented on src/main/kotlin/blackjack/view/BlackJackInputView.kt:
```diff
@@ -0,0 +1,9 @@
+package blackjack.view
+
+interface BlackJackInputView {
```
이전 코멘트에 남겨두었던 interface로 하셨군요💪
하시면서, 어떤 생각이 드셨을까요?
굳이 이렇게까지 해야하나? 라는 생각도 좋고 해당 구조로 하면 어떻게 활용해볼 수도 있겠다등 어떠한 의견이라도 좋습니다.

moondev03 replied:
지금은 굳이 이렇게까지 해야하나? 라는 생각이 들기도 하지만,
BlackJack View Interface를 작성하므로써 View 구현체에 따라 사용자가 원하는 다양한 형태의 뷰를 즐길 수 있을 것 같습니다.

Gyuil-Hwnag commented on src/main/kotlin/blackjack/view/ViewExtension.kt:
```diff
@@ -0,0 +1,36 @@
+import blackjack.domain.state.PersonState
+import java.text.DecimalFormat
+
+fun Double.formatAmount(): String {
+    val formatter = DecimalFormat("#.##")
+    return formatter.format(this)
+}
```
Extensions로 분리를 해주셨군요👍
View에서의 사용을 명확하게 알리기 위해서 ViewExtension이라는 네이밍으로 지어주셨나 보네요

Gyuil-Hwnag reviewed (changes requested):
다이스 안녕하세요!
해당 미션 구현 고생 많으셨습니다!
미션 구현을 너무 잘 해주셨습니다 💯

고민해보시면 좋을만한 내용들 코멘트를 달았으니 확인해주시면 감사하겠습니다.
이전 미션에서 너무 잘 구현해주셔서, 내용이 많지는 않았네요.
이번 PR 올리기 전에 다이스가 구현하신 내용을 살펴보고, 변경해보고 싶은 내용이 있다면 어떤 것이든 변경하고 의견을 남겨주세요.
(이전에 코멘트로 수정했던 내용에 대한 롤백도 환영입니다.)

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
화이팅입니다💪

moondev03 committed:
test: BlackJack인 경우 테스트 실패 수정

- game.dealCards를 사용하지 않고 바로 플레이어 턴 수행하도록 변경

moondev03 committed:
refactor: GameResult 개선

- 파일 위치 변경
- Lose 케이스 함수 분리
- totalDealerProfit 초기값 상수화

moondev03 committed:
feat: Deck 리필 기능 구현

moondev03 committed:
fix: 수익률 계산 결과 수정

- 블랙잭으로 승리 -> 1.5배
- 승리 -> 1배
- 패배 -> 배팅금 잃음
- 무승부 -> 배팅금 반환

moondev03 committed:
refactor: 딜러와 플레이어 모두 버스트가 아닐 때 플레이어만 블랙잭인 케이스 추가

moondev03 commented:
## 추가 및 수정된 기능

### 카드 리필 기능
인원 수를 제한하지 않는 대신 미리 만들어둔 하나의 카드 목록을 계속 재사용하는 식으로 Deck을 리필하도록 했습니다.

```Kotlin
companion object {
    private const val DECK_SIZE = 52
    private const val INVALID_DECK_SIZE_ERROR_MESSAGE = "덱은 52장의 카드로 구성되어야 합니다. (%s 장)"

    private val CACHED_CARDS: List<Card> = CardPattern.entries.flatMap(::createCard)
        get() = field.shuffled()

    private fun createCard(cardPattern: CardPattern): List<Card> = CardNumber.entries.map { Card.create(it, cardPattern) }
}
```

<br>

### 수익률 계산 수정
딜러와 플레이어 수익률 계산 방법을 착각하여 다시 수정했습니다.

**플레이어 수익**:
- 블랙잭으로 승리 -> 1.5배
- 승리 -> 1배
- 패배 -> 배팅금 잃음
- 무승부 -> 배팅금 반환 (수익 0)


**딜러 수익**:
`Lose인 플레이어의 배팅금 총합`에서
`플레이어 수익 x -1`로 변경

<br>

### 딜러와 플레이어 모두 21일 때 플레이어만 블랙잭인 케이스 추가

Gyuil-Hwnag commented on src/main/kotlin/blackjack/controller/BlackJackController.kt:
```diff
@@ -1,14 +1,15 @@
 
 class BlackJackController(
-    private val inputView: InputView = InputView(),
-    private val outputView: OutputView = OutputView(),
+    private val inputView: BlackJackInputView,
+    private val outputView: BlackJackOutputView,
 ) {
```
너무 잘해주셔서, 코멘트 내용이 많지 않을 것 같은데, 시간이 있는 만큼 콜백 형태로도 구현을 해봤으니, 함수의 파라미터로 받아서 처리를 하는 방법도 구현을 해보고 두개를 비교해보는 것 과정에서도 얻을 것이 있을 것 같다는 생각이 듭니다💪
Controller 에서 `BlackjackGame.play()` 하나로 블랙잭 게임을 할 수 있는 형태로요

Gyuil-Hwnag commented on src/main/kotlin/blackjack/controller/BlackJackController.kt:
```diff
@@ -49,6 +64,16 @@ class BlackJackController(
+        outputView.printGameResult(gameResult)
+    }
+
+    private fun <T> retryWhenException(action: () -> T): T {
+        while (true) {
+            runCatching {
+                return action()
+            }.onFailure { e ->
+                outputView.printMessage(e.message)
+            }
+        }
```
해당 내용은 실제 프로젝트에서 사용한다면 utils or core등으로 파일을 분리해서, 사용할 수 있도록 하겠군요!

Gyuil-Hwnag commented on src/main/kotlin/blackjack/domain/person/Player.kt:
```diff
@@ -1,8 +1,12 @@
-class Player(val name: String) : Person() {
+class Player(
+    val name: String,
+    val betAmount: BetAmount = BetAmount(),
+) : Person() {
```
개인적으로 생성자에 기본 생성자가 있는건 지양을 하는편이긴 합니다.
(이건 물론 사람마다 방식마다 다를 수 있겠지만요)
얼마전에 회사 코드에서 기본생성자를 넣었다가 에러를 발생시켰던 일이 있어서 지양을 하게 되었던 것 같아요

기본 생성자로 되어 있다보면, 개발자가 자칫 잘못해서 생성시에 빠트리는 경우가 있습니다.
영향력이 적거나 없는 경우라면 괜찮겠지만, 예를 들어서 물건의 가격등이라면 아주 큰 문제가 발생할 수 있을 것 같네요🥲

moondev03 replied:
기본 생성자 제거해봤습니다!

4a94b1b068c0f85cdc0e9a72ba3e2fc4661f8391

Gyuil-Hwnag commented on src/main/kotlin/blackjack/domain/card/Deck.kt:
```diff
@@ -1,28 +1,32 @@
-
-    private fun createCard(cardPattern: CardPattern): List<Card> {
-        return CardNumber.entries.map { cardNumber -> Card.create(cardNumber, cardPattern) }
+    private fun regenerateDeck() {
+        _cards.addAll(CACHED_CARDS)
+        validateDeckSize()
     }
```
덱이 다 떨어졌을 때, 리필을 하는 기능 추가 좋습니다👍

Gyuil-Hwnag commented on src/main/kotlin/blackjack/domain/result/GameResult.kt:
```diff
@@ -0,0 +1,56 @@
+import blackjack.domain.state.PersonState
+import blackjack.domain.state.ResultState
+
+class GameResult(dealer: Dealer, players: List<Player>) {
+    val playerPayouts: Map<Player, Double>
+    val dealerProfit: Double
```
수익률 계산에 대해서 이전에 잘못들어가 있던 내용들이 변경이 되었군요👍
돌라보다가, 뭔가 이상해서 이번 리뷰에 남기려고 했었는데 아주 좋습니다

Gyuil-Hwnag reviewed (changes requested):
다이스 안녕하세요!

해당 미션 리뷰 반영 고생 많으셨습니다!
리뷰 반영하면서 구현한 내용 수정 너무 잘 해주셨습니다 💯

구현과 반영 너무 잘해주셨습니다👍
시간이 있는 만큼 남겨둔 코멘트 처럼 여러 방법으로도 구현을 해보고 각 장단점을 비교해보면 얻어갈 수 있는 것들이 있을 것 같다는 생각이 드네요
참고하면 좋을만한 내용들 코멘트를 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
화이팅입니다💪

moondev03 committed:
refactor: BetAmount 초기값 제거 및 불필요한 메서드 제거

moondev03 committed:
feat: Profit Value Class 추가

moondev03 committed:
refactor: 예외 발생 시 재실행 함수 Controller에서 분리

moondev03 committed:
refactor: PersonCards 제거

- Hand에서 카드 관리
- 기존 Score는 Person에서 반환

moondev03 committed:
refactor: Score 객체 생성 방식 변경

- factory method
- PersonState에 있던 Bust, DealerStay, BlackJack 점수 판단을 Score로 이전
- 테스트 코드 추가 및 수정

moondev03 committed:
refactor: BlackJackGame 개선

- View Interface를 주입하도록 변경

moondev03 committed:
refactor: Player에서 배팅 정보 분리

- Player와 BetAmount를 감싸는 PlayerBetInfo 객체 생성
- GameResult에서 승패에 따른 Profit 계산 로직 분리

moondev03 committed:
test: BlackJackGameTest 작성

moondev03 committed:
refactor: Score Private Constructor로 수정

moondev03 commented:
### PlayerBetInfo
`Player`에서 관리하던 배팅 금액을 `PlayerBetInfo` 객체를 만들어 관리하도록 수정했습니다.
해당 객체에서는 기존 `GameResult`에서 수행하던 승패에 따른 배당금 계산을 수행하도록 했습니다.

`PlayerBetInfo`: 각 플레이어에 대한 배팅 정보를 관리하고, 게임의 결과에 따른 플레이어의 배당금을 계산하는 역할을 맡습니다.

`GameResult`: 게임의 최종 결과를 집계하며, 각 플레이어와 딜러의 수익을 계산하고 관리하는 역할을 합니다.

<br>

### BlackJackGame
`Controller` 에서 `BlackjackGame.play()` 하나로 블랙잭 게임을 할 수 있는 형태로 수정했습니다.

InputView/OutputView <- BlackJack View Interface <- BlackJackGame

해당 구조는 추상화된 View 구현체를 통해 다양한 컨트롤러에서 사용될 수 있다는 장점이 있는 것 같습니다!

Gyuil-Hwnag commented on src/main/kotlin/blackjack/domain/BlackJackGame.kt:
```diff
@@ -3,65 +3,117 @@ package blackjack.domain
 class BlackJackGame(
-    val dealer: Dealer,
-    players: List<Player>,
-    private val deck: Deck = Deck(),
+    private val inputView: BlackJackInputView,
+    private val outputView: BlackJackOutputView,
 ) {
```
BlackJackGame을 이전 callback 형태에서, callback을 제거하는 방식으로 잘 변경하셨군요👍
이번 작업을 하면서 다이스가 느끼신 점들이 있다면 알려주세요~
그 어떤 것이라도 좋습니다!

moondev03 replied:
콜백으로 인해 컨트롤러에 중복 코드가 제거되니 코드가 깔끔해진 것 같습니다!

추가로 의존성 역전 원칙(DIP) 개념에 대해 찾아볼 수 있는 기회가 된 것 같습니다!

Gyuil-Hwnag commented on src/main/kotlin/blackjack/domain/Score.kt:
```diff
@@ -3,43 +3,39 @@ package blackjack.domain
-class Score(cards: List<Card>) {
-    private var _value: Int = calculate(cards)
-    val value: Int get() = _value
+class Score private constructor(val value: Int) {
+    fun isBlackJackScore(): Boolean = value == BLACKJACK_SCORE
```
Score객체에 대한 리팩토링이 이루어졌군요~👍

Gyuil-Hwnag commented on src/main/kotlin/blackjack/domain/person/Person.kt:
```diff
@@ -10,14 +11,20 @@ abstract class Person {
 
     private val hand: Hand = Hand()
```
아래와 같이 사용을 하는 방법도 있습니다~
다이스가 보시기에는 각 두방식은 어떤 차이점이 있을까요?
~~~kotlin
    protected var state: PersonState = PersonState.HIT
    val gameState: PersonState get() = state
    
    // 예시
    var state: PersonState = PersonState.HIT 
        private set
~~~

moondev03 replied:
Person을 상속받는 Player에서는 STAY 상태로의 전환을 허용해야 해서 해당 형태로 작성했습니다.

원래는 Person에서 상태를 변경하는 함수를 만들었지만, 
STAY 상태로 변경하는 기능은 Player에서만 필요했기 때문에 이런 구조가 된 것 같습니다.

해당 내용 더 고민해보고 수정해보겠습니다!

Gyuil-Hwnag commented on src/main/kotlin/blackjack/domain/person/PlayerBetInfo.kt:
```diff
@@ -0,0 +1,45 @@
+import blackjack.domain.state.PersonState
+import blackjack.domain.state.ResultState
+
+class PlayerBetInfo(
+    val player: Player,
+    private val betAmount: BetAmount,
+) {
```
PlayerBetInfo 객체를 통해서, 플레이어가 이기고 졌을때 등에 대한 게임의 배팅 결과를 알려주는 군요👍

Gyuil-Hwnag commented on src/main/kotlin/blackjack/domain/state/PersonState.kt:
```diff
@@ -1,25 +1,26 @@
 import blackjack.domain.person.Dealer
 import blackjack.domain.person.Person
 
 enum class PersonState(val isFinal: Boolean) {
     HIT(false),
```
이제 isFinal 이라는 변수는 따로 사용이 안되고 있군요~

moondev03 replied:
```kotlin
abstract class Person {
...
    fun isDrawable(): Boolean {
        return !state.isFinal
    }
...

```

Person에 있는 isDrawable()에서 사용하고 있습니다!

Gyuil-Hwnag commented on src/test/kotlin/blackjack/domain/BetAmountTest.kt:
```diff
@@ -0,0 +1,38 @@
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+
+class BetAmountTest {
+    private lateinit var betAmount: BetAmount
```
BetAmount에 대한 테스트 좋습니다~👍

Gyuil-Hwnag commented on src/test/kotlin/blackjack/domain/BlackJackGameTest.kt:
```diff
@@ -1,62 +1,70 @@
 class BlackJackGameTest {
-    private lateinit var dealer: Dealer
-    private lateinit var players: List<Player>
-    private lateinit var game: BlackJackGame
+    private lateinit var outputStream: ByteArrayOutputStream
+    private lateinit var printStream: PrintStream
```
Controller 테스트 형태를 작성하셨군요👍

Gyuil-Hwnag commented on src/test/kotlin/blackjack/domain/BlackJackGameTest.kt:
```diff
@@ -1,62 +1,70 @@
-    fun `딜러와 플레이어가 각각 2장의 카드를 받아야 한다`() {
-        game.dealCards()
+    fun `블랙잭 게임을 정상적으로 수행한다`() {
+        setInput("player1, player2\n100\n100\nn\nn\n")
+
+        BlackJackGame(InputView(), OutputView()).play()
+
+        val output = outputStream.toString()
 
         assertAll(
-            { dealer.cards.size shouldBe 2 },
-            { players.all { it.cards.size == 2 } shouldBe true },
+            { assertTrue(output.contains("게임에 참여할 사람의 이름을 입력하세요")) },
+            { assertTrue(output.contains("배팅 금액은?")) },
+            { assertTrue(output.contains("한장의 카드를 더 받겠습니까?")) },
+            { assertTrue(output.contains("## 최종 수익")) },
         )
     }
```
Ui 테스트 형태를 작성해보셨군요👍

Gyuil-Hwnag commented on src/test/kotlin/blackjack/domain/person/PlayerBetInfoTest.kt:
```diff
@@ -0,0 +1,85 @@
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+
+class PlayerBetInfoTest {
+    private lateinit var playerBetInfo: PlayerBetInfo
+    private lateinit var dealer: Dealer
```
Player 배팅 결과에 대한 테스트 좋습니다!

Gyuil-Hwnag reviewed (approved):
다이스 안녕하세요~
미션 정말 고생 많으셨습니다! 👏

다음 미션 진행을 위하여 이번 미션은 여기서 머지하도록 하겠습니다.

구현을 하시면서 여러가지 고민도 많이 하시고 나왔던 키워드들이 조금씩 있었는데, 해당 내용들 잘 적용해서 다음미션도 잘 진행 하시리라 믿고 응원하겠습니다😆
이번 미션을 진행하시면서 감기 때문에 몸도 안좋으셨을텐데, 정말 고생 많으셨어요..
미션도 중요하긴 하지만, 어디까지나 몸 건강 챙기는게 우선입니다!💪
이번 미션 정말 너무나 고생 많으셨어요💯

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
앞으로도 응원하겠습니다 💪
